pkg/app/piped/appconfigreporter/appconfigreporter.go

```diff
@@ -0,0 +1,371 @@
+		return nil, fmt.Errorf("failed to get files those were touched between two commits: %w", err)
+	}
+	if len(files) == 0 {
+		// The case where all changes have been fully reverted.
+		return []*model.ApplicationInfo{}, nil
+	}
+	apps := make([]*model.ApplicationInfo, 0)
+	for _, filename := range files {
+		if !strings.HasSuffix(filename, model.DefaultDeploymentConfigFileExtension) {
+			continue
+		}
+		gitPathKey := makeGitPathKey(repoID, filename)
+		if _, registered := registeredAppPaths[gitPathKey]; !registered {
+			continue
+		}
+		appInfo, err := r.readApplicationInfo(ctx, filepath.Join(repo.GetPath(), filename))
+		if err != nil {
+			r.logger.Error("failed to read application info",
+				zap.String("repo-id", repoID),
+				zap.String("config-file-path", filename),
+				zap.Error(err),
+			)
+			continue
+		}
+		apps = append(apps, appInfo)
+	}
+	return apps, nil
+}
+
+// updateUnregisteredApps sends all unregistered application configurations to the control-plane.
+func (r *Reporter) updateUnregisteredApps(ctx context.Context, registeredAppPaths map[string]struct{}) error {
+	apps := make([]*model.ApplicationInfo, 0)
+	for repoID, repo := range r.gitRepos {
+		as, err := r.findUnregisteredApps(ctx, repo.GetPath(), repoID, registeredAppPaths)
+		if err != nil {
+			return err
+		}
+		r.logger.Info(fmt.Sprintf("found out %d unregistered applications in repository %s", len(as), repoID))
+		apps = append(apps, as...)
+	}
+	if len(apps) == 0 {
+		if r.sweptUnregisteredApps {
+			return nil
+		}
+		r.sweptUnregisteredApps = true
+	} else {
+		r.sweptUnregisteredApps = false
+	}
+
+	_, err := r.apiClient.ReportUnregisteredApplicationConfigurations(
+		ctx,
+		&pipedservice.ReportUnregisteredApplicationConfigurationsRequest{
+			Applications: apps,
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("failed to put the latest unregistered application configurations: %w", err)
+	}
+	return nil
+}
+
+// findUnregisteredApps finds out unregistered application info in the given git repository.
+func (r *Reporter) findUnregisteredApps(ctx context.Context, repoPath, repoID string, registeredAppPaths map[string]struct{}) ([]*model.ApplicationInfo, error) {
+	return r.scanAllFiles(ctx, repoPath, repoID, registeredAppPaths, false)
+}
+
+// scanAllFiles inspects all files under the root or the given repository.
+// And gives back all application info as much as possible.
+func (r *Reporter) scanAllFiles(ctx context.Context, repoRoot, repoID string, registeredAppPaths map[string]struct{}, wantRegistered bool) ([]*model.ApplicationInfo, error) {
+	apps := make([]*model.ApplicationInfo, 0)
+	err := fs.WalkDir(r.fileSystem, repoRoot, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() {
+			return nil
+		}
+
+		cfgRelPath, err := filepath.Rel(repoRoot, path)
+		if err != nil {
+			return err
+		}
+
+		if !strings.HasSuffix(cfgRelPath, model.DefaultDeploymentConfigFileExtension) {
+			return nil
+		}
```

Comment:
nit: We can move this to before line 305 and use `path` to check the extension. It will reduce a lot of unnecessary `filepath.Rel` calls.

Reply:
Good catch